docs: add API documentation

ca-d opened the pull request:
This is our initial draft of an API documentation.

The document is aspirational in nature, meaning it does not reflect the current state of the API, but rather where we would like to be. Code implementing this API is already available in separate branches, so once we have accepted a version of this API doc we can immediately bring the implementation up to speed.

danyill commented on API.md:
````diff
@@ -0,0 +1,319 @@
+}
+```
+
+It's `title` property is a human-readable description of the edit, and `squash` is a boolean indicating whether the edit should be merged with the previous edit in the history.
````
```suggestion
It's `title` property is a human-readable description of the edit (for display in editing history), and `squash` is a boolean indicating whether the edit should be merged with the previous edit in the history.
```

Do I understand the purpose correctly that the title is mainly for display in editing history?

JakobVogelsang replied:
Yes. At least that is how I did understand this. It makes sense mostly with complex edits like the importing of an IED, where - if you want to display a history to the user - it would be beneficial to have a better description of the complex edit in the hostory.

ca-d replied:
No other app that I know of even displays an "edit history". Microsoft Word, e.g., only shows "undo" and "redo" buttons, respectively, and doesn't tell you what it is that would be undone or redone.

If we were to remove this "title" feature and the "edit history" view, we'd have a simpler API and avoid the problem of history entries not being translatable once they have been recorded. I am in favour of this approach.

What do you think, @danyill and @JakobVogelsang ?

JakobVogelsang replied:
I am torn about that. I am using it a lot at the moment, BUT if I am being honest, that is more for debugging while plugin development than anything else. Other users have reported a different story.

That said, I still would provide the history, because other plugin might want to better display the history or do something like writing to the `History` item. I would also still have the `sqash` key now in order to have correct redo states, which is a bit tricky sometimes at the moment.

ca-d replied:
Of course, the `squash` flag is needed for undo and redo to not infuriate end users. Whether the `title` is really needed is still an open question, though. For `Hitem`s a hand-written piece of rich text by the end user that describes a whole bunch of related edits (much like a `git` commit message) seems more sensible to me. I'd love to have more feedback from actual users (i.e. engineers) on this.

danyill replied:
> No other app that I know of even displays an "edit history". Microsoft Word, e.g., only shows "undo" and "redo" buttons, respectively, and doesn't tell you what it is that would be undone or redone.

Oh but it does! However over the undo and redo buttons some time ;-) I remember it from before the ribbon days when it was suffixed to the words "Undo" and "Redo" in the drop-down menu.

danyill replied:
> Whether the title is really needed is still an open question, though.

I agree it's an open question but Microsoft Word and others do have a similar feature.
They also allow you to multiselect a number of changes down to a certain level to undo them which I do think is rather nice (and not a feature I use very often):

![image](https://github.com/user-attachments/assets/a4852b2d-a7f5-4550-873b-a3bb4070db93)

Interestingly the same menu is not maintained for the "Redo" function.

danyill replied:
My feeling is, "If in doubt we should add it". I like Jakob's point about it being useful for development and I also think users often want an assurance on a step of actions recently taken (particular if they've been clicking a lot of buttons with not directly visible consequences). Being able to see a drop down even just to see "what I did recently" has some utility.

ca-d replied:
> > No other app that I know of even displays an "edit history". Microsoft Word, e.g., only shows "undo" and "redo" buttons, respectively, and doesn't tell you what it is that would be undone or redone.
> 
> Oh but it does! However over the undo and redo buttons some time ;-) I remember it from before the ribbon days when it was suffixed to the words "Undo" and "Redo" in the drop-down menu.

Thanks for the lesson! It really shows that I don't use Microsoft software day to day.

> My feeling is, "If in doubt we should add it".

Generally, my feeling is "If in doubt, leave it out!", but in this particular case I can see the usefulness. I also always wanted to have the feature where if you click on a non-current edit, you jump to the state of the document just after that edit. If you click on the current edit, you "undo". That way any place in the edit history is at most a double-click away. Would be fairly trivial to implement in core, but I haven't found the time for that nice-to-have yet.

danyill commented on API.md:
````diff
@@ -0,0 +1,319 @@
+export type Edit = Insert | SetAttributes | Remove | Edit[];
+```
+
+This means that a single edit can either consist in a sequence of other edits or in one of the following atomic edit types:
````
It's a bit of an edge case, but changing `textContent` is quite awkward at the moment.
Is it worth having an API for this? I am happy if you decide not to.

JakobVogelsang replied:
Would you then expect a new edit type `SetTextContent`?

danyill replied:
Yes please.

ca-d replied:
Great point, I'll add this to the proposal.

danyill commented on API.md:
```diff
@@ -0,0 +1,319 @@
+export type SetAttributes = {
+  element: Element;
+  attributes: Partial<Record<string, string | null>>;
+  attributesNS: Partial<Record<string, Partial<Record<string, string | null>>>>;
```
For context: https://github.com/openscd/open-scd-core/issues/132

danyill commented on API.md:
```diff
@@ -0,0 +1,319 @@
+
+Currently, the only way to open an editor plugin is by clicking on its tab in the tab bar. It would be nice to be able to deep link to a particular editor plugin, for example by clicking on a link in another plugin.
+
+This could be handled by the OpenSCD distribution, which could e.g. listen for `hashchange` events and open the appropriate editor plugin by setting the OpenSCD core's `editorIndex` property.
```
```suggestion
This could be handled by the OpenSCD distribution, which could e.g. listen for `haschange` events and open the appropriate editor plugin by setting the OpenSCD core's `editorIndex` property.
```

Someone's got hashes in focus ;-)

More seriously do we have an approach for translations and changing of locale for open-scd-core that works? Or is this just a plugin concern?

ca-d replied:
I don't currently see any other way to do translations (across arbitrary plugins implemented using different technology stacks) than to just hand the currently selected language code to the plugin and leave translation to the plugin author.

With English as a default language by convention, I expect most plugin authors will chose not to translate their plugins at all.

Since users (even from non-anglophone countries) have hitherto unanimously expressed no desire for localization of the app, another possibility would be to remove localization from OpenSCD core altogether. This would be uncomfortable for OMICRON (whose customers in some regions traditionally expect fully localized products), but might be easier on the community.

What do you think about this, @danyill and @JakobVogelsang ?

JakobVogelsang replied:
I would not get rid of the `locale` property. Whether to have it in a plugin or even in the core component is an individual choice.

ca-d replied:
Alright, let's leave it in.

danyill commented on API.md:
```diff
@@ -0,0 +1,319 @@
+export type EditDetailV2<E extends Edit = Edit> = {
+  edit: E;
+  title?: string;
+  squash?: boolean;
```
On contemplation this does seem like an elegant way to resolve the poorly worded concerns ([here](https://github.com/OMICRONEnergy/oscd-designer/issues/57)) without introducing new problems.

What does "squash" do? Is it purely a "grouping after the event" (literally).
What if a plugin (e.g. a menu plugin) listens to another plugin and pushes out events in response to an EditEvent. 

Would this result in these edits being squashed or handled in an unexpected fashion? I'm particularly thinking of what I do in the subscriber plugin. I register a `window` event ([here](https://github.com/danyill/oscd-subscriber-later-binding/blob/670598f6cc5c64168f3e9b59eef8c7e03eadf886/oscd-subscriber-later-binding.ts#L546-L557 )) and then the IED specific plugins generate new events if appropriate ([here](https://github.com/danyill/oscd-subscriber-lb-siemens/blob/ae20db2895c4c9f73029d090d3fae0019ae51854/oscd-subscriber-lb-siemens.ts#L217-L227 ), it may not be done well!). I guess in this case I would like the dependent events to be included in the initiating event but perhaps this wouldn't always be the case?

JakobVogelsang replied:
Whenever a PR is coming with the `squash` flag, the editing handle does merge it with the edit before. This would allow a plugin author to avoid unwanted redo states. For your example, you could use the flag to sqash the edit coming from the additional menu plugins with the ones that triggered the menu plugin. 

But I agree in general. The event handling in the browser does allow you to listen to any events and send your own events. We therefore cannot guarantee the there is not an in between edit event coming in and spoling the sqash.

danyill replied:
Thinking about it I feel that plugins and dependent plugins can sort out their own in-house rules and if they spoil the squash, it's on them. They can use the composed path to access information from the event source like I do now:

```js
function shouldListen(event: Event): boolean {
  const initiatingTarget = <Element>event.composedPath()[0];
  return (
    initiatingTarget instanceof Element &&
    initiatingTarget.getAttribute('identity') ===
      'danyill.oscd-subscriber-later-binding' &&
    initiatingTarget.hasAttribute('allowexternalplugins')
  );
}
```

So I think no change or improvement is required.

ca-d replied:
Thank you both for the discussion. My takeaway is that the behaviour of the `squash` flag and its use cases need to be made very clear in the eventual API documentation (generated from code).

danyill reviewed (approved):
I am delighted to see further API development and pleased to see this timely proposal.
LGTM.

JakobVogelsang commented on API.md:
```diff
@@ -0,0 +1,319 @@
+
+It is expected that the fonts `--oscd-theme-text-font` and `--oscd-theme-icon-font` will be loaded in OpenSCD's `index.html` file. OpenSCD core does not load any fonts by itself.
+
+## Missing
```
I am not entirely sure, if we should have this chapter. It seems weird to read this in an API document. We define tally should have these ideas saved in a very prominent way, but the API document seems to not be a good place.

danyill replied:
Perhaps the relevant point is "Why we don't need to expand or modify this API for these particular concerns". But by all means we could split this off into its own document?

ca-d replied:
My intention was to obsolesce the "Missing" chapter by discussing all points here, and then delete it before merging.

danyill commented on API.md:
````diff
@@ -0,0 +1,319 @@
+
+The **open event** allows a plugin to add a document `doc` to the `docs` collection under the name `docName`.
+
+```typescript=
````
```suggestion
```typescript
```

danyill commented on API.md:
````diff
@@ -0,0 +1,319 @@
+
+The **wizard event** allows the plugin to request opening a modal dialog enabling the user to edit an arbitrary SCL `element`, regardless of how the dialog for editing this particular type of element looks and works.
+
+```typescript=
````
```suggestion
```typescript
```

danyill reviewed (commented):
A few more thoughts and perhaps we could get some syntax highlighting fix.

danyill commented on API.md:
```diff
@@ -0,0 +1,319 @@
+
+### `EditEventV2`
+
+The **edit event** allows a plugin to describe the changes it wants to make to the current `doc`.
```
Last comment I promise! 

Is an edit event sometimes unsuccessful? Should our API allow a positive assertion of success/failure, like an exit code? Or is this unnecessary? (similar questions around our library development in scl-lib have been discussed recently and I would be interested in thoughts on this).

ca-d replied:
Sadly, [dispatching an event only returns whether the event was cancelled, according to the Web API](https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/dispatchEvent#return_value). So the only feedback a plugin gets for dispatching an event is a change (or no change) to the edit `history`, which as of this API draft would be fully exposed to plugins through a property.

I'm open to other suggestions for convenient feedback mechanisms. Simply adding an optional callback to the event which will be called by core upon completion of the edit (successful or otherwise) along with some status code seems a bit ugly and cumbersome, but would be a last resort stop-gap measure for use cases for which the above updating of the `history` property would not suffice.

What are the use cases that you see for this feedback mechanism, @danyill and @JakobVogelsang ?

JakobVogelsang replied:
For me it is again mainly when developing plugins. You simply do not know if the code is working correct, and then feedback mechanisms are important. A good indicator could be feedback on the edits send. I am not sure though if it is needed in production. Also I think this is something a plugin author could take care as well. When edits are sent it is quite predictable what the core component will do.

danyill replied:
I tend to agree, if a plugin sends through sane edits then core can generally be relied on. 

Then it becomes the task of users of core to send through correct edits in the first place. So perhaps an "exit code" is more a requirement on a downstream library or the responsibility of an individual plugin.

For the DOM API I note that `insertBefore` returns the inserted node whereas `setAttribute` doesn't return anything but can have exceptions which can be managed.

Should plugins be able to listen to an event to indicate that something has gone wrong during the edit process? (either the node not being inserted or an exception occurring)? I'm thinking something to do with the catch block: 

https://github.com/OpenEnergyTools/open-scd-core/blob/3f9aacb94eec216d6a27b9407fa1dc06714950ea/foundation/edit-event.ts#L73-L92

I am not knowledgeable about API design and error handling so I'll be happy for this comment to be passed over 😉

ca-d replied:
Hope this is good enough for our use cases. If not, we can open a further issue to extend `Edit`s with return-code callbacks or the like at a later point.

ca-d commented on API.md:
```diff
@@ -0,0 +1,319 @@
+export type EditDetailV2<E extends Edit = Edit> = {
+  edit: E;
+  title?: string;
+  squash?: boolean;
```
> I suppose what I'd end up doing with API is something like this which seems OK

Another possibility would be:
> ```ts
> let squashable = false
> doSomeStuff.forEach(stuff => {
>   // blah blah
>   const someEdit = doCoolStuff(stuff)
> 
>   this.dispatchEvent(
>       newEditEvent2(
>           someEdit,
>           title: "Look Ma a Title",
>           squash: squashable 
>         )
>       );
>   squashable = true
> })
> ```

ca-d commented on API.md:
```diff
@@ -0,0 +1,319 @@
+
+It is expected that the fonts `--oscd-theme-text-font` and `--oscd-theme-icon-font` will be loaded in OpenSCD's `index.html` file. OpenSCD core does not load any fonts by itself.
+
+## Missing
```
The proposal in [this comment](https://github.com/OpenEnergyTools/scl-lib/issues/54#issuecomment-2031855291) is something we could include in the API defined here. What do you think about that?

To me it looks like a great fit for the new `SetAttributes` API, since we already provide automatically generated unique namespace prefixes there (as opposed to the old `Update` API that left prefixes to the user). Collating all namespace definitions for newly inserted elements and newly set attributes into the root node is quite a bit of extra effort on the API implementor's side, but would make the end user's experience much smoother.

It would also be the first bit of logic in OpenSCD core that is `SCL` specific. At that point, we could also consider reintroducing built-in `getReference` (for `reference === undefined` in `Insert`) by letting OpenSCD core depend on `scl-lib`.

What do you think about these points, @danyill and @JakobVogelsang ?

JakobVogelsang replied:
I would love to have it in. That is from what I understand an implementation issue in the core component. The proposed API allows to do that, right?

danyill replied:
I like it. It offloads responsibility onto core for some slightly complicated semantics and helps keep the SCL documents consistent with the API and that's a good "core task".

ca-d replied:
Could one of you send a pull request onto my `add-api-docs` branch with a draft of the required API changes? If possible, including the automatic `reference` if undefined, since that makes keeping the document at least schema-valid a much easier task for plugin authors who don't want to think about the standard too much.

ca-d replied:
We've decided in today's meeting to not implement those features within core, but to leave `getReference` in `scl-lib` and add a functional API for automatic handling of extension namespaces there instead.

ca-d commented on API.md:
````diff
@@ -0,0 +1,319 @@
+OpenSCD core communicates the data necessary for editing SCL documents by setting the following [properties](https://developer.mozilla.org/en-US/docs/Glossary/Property/JavaScript) on the plugin's [DOM Element](https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement):
+
+
+```typescript=
````
Thanks for pointing this out! We originally drafted this in [HedgeDoc](https://hedgedoc.org/), which doesn't support `ts` (only `typescript`). I'll remove the `=` signs.

ca-d commented:
@danyill I also just added your "plugin bundle archive format" suggestion to the "Missing" section at the bottom.

Like all features which I'd rather implement in a distribution or within a plugin, I added it for completeness of our discussion only, and would suggest removing it from `API.md` before merging, as soon as we're all agreed that it doesn't belong in core.

ca-d commented:
I just removed the "Missing" section from the bottom. If you think any of the points it listed need to be handled by OpenSCD core in particular, please discuss here:

## Missing

### Plugin manifest, searchable plugin store

This is still needed if we want to enable plugins to be installed and updated from within OpenSCD. This could be done by a "plugin management" or "plugin store" plugin. Otherwise, this could be done by the OpenSCD distribution itself.

A good candidate for a plugin manifest format is the Plugin type defined above, with the addition of a `kind` flag indicating whether it is a `"menu"` or `"editor"` plugin (or maybe `"both"`), and possibly a `version` property.

### Plugin bundle archive format

For offline installation of plugins (e.g. by dragging and dropping a `.zip` file into OpenSCD), a format for bundling the plugin's JavaScript and resource files along with a manifest file is needed. This could either be implemented by the OpenSCD distribution or by a "plugin management" plugin as described in the previous section.

### Plugin update notifications

Given a plugin manifest with a `version` number, OpenSCD core could check for updates to the plugins and notify the user when updates are available.

Within the current architecture, the plugin's own service worker is expected to handle this.

### Deep linking to editor plugins

Currently, the only way to open an editor plugin is by clicking on its tab in the tab bar. It would be nice to be able to deep link to a particular editor plugin, for example by clicking on a link in another plugin.

This could be handled by the OpenSCD distribution, which could e.g. listen for `hashchange` events and open the appropriate editor plugin by setting the OpenSCD core's `editorIndex` property.

trusz commented on API.md:
```diff
@@ -0,0 +1,329 @@
+
+Its `title` property is a human-readable description of the edit for displaying in the editing history dialog.
+
+The `squash` flag indicates whether the edit should be merged with the previous edit in the history.
```
I am not sure if I get this correctly so here is an example:
```ts
const edit0 = { title: "some meaningful message", squash: false, ... }
// from this point forward, I don't need to a title because it is going to be squashed with the previous one
const edit1 = { title: undefined, squash: true, ... }
const edit1 = { title: undefined, squash: true, ... }
```

what happens if I send `squash:true` and so smuggle my change into a history item that is not mine?

```ts
const editFromPluginA = { title:"some uncritical change", ... }

const editFromPluginB = { title:"I will be part of the previous history item", squash:true }
```

another edge case would be if I send the first change with `squash:true` and there is no item to squash with. I assume we should just silently not do anything.


My first thought was that if I want to group changes into a single history item we could send an array of changes with a title. Something along the lines of:

```ts
const batchChanges = {
  title:"I've made a lots of changes",
  squash: true.
  events: [
   { ... },
   { ... },
   { ... },
  ]
}
```
In this case `squash` would be also redundant maybe.

However as I understand some event have dependencies on each other. e.g.:

```ts
const addSomeItemEvent = { title:"added flux capacitor", squash: false }
const initFluxCapacitorCapacitiy = { title: undefined, squash: true }
```
so I (maybe) cannot batch these together because the element in the second event does not exist and I have to wait for it.
Do I get this right?


The concept of "squashing with previous" feels somewhat error-prone. Would be "squash with this ID" too much?
e.g:
```ts
const aRandomId = generateID()
const addSomeItemEvent = { title:"added flux capacitor", id: aRandomId }
const initFluxCapacitorCapacitiy = { title: undefined, squashWith: aRandomId }
```

ca-d replied:
"squash with ID" would actually turn out more problematic in terms of edge-cases, our undo/redo history really has to be completely linear and in sync with the clock time at which the edit is composed. This sad circumstance is directly due to us using the *mutable* `DOM` API for data exchange, referencing live `Node`s and `Element`s in our `Edit`s.

The two edge cases you mention are straightforward to resolve:
- If the `history` is empty, a new `Edit` is created as if `squash` was `false`.
- If the last `Edit` was dispatched by a different plugin, we still allow `squash`ing my own edit onto it.
  This allows for a real improvement of the user experience in certain scenarios. @danyill 's manufacturer-specific "automatic supervision" plugins, for example, might prefer `squash`ing onto the previous edit, their edits thereby being un-done and re-done at the same time as the subscription they are supervising. This would also enable completely new use-cases like "automatic reference update" or "automatic cleanup" plugins that remove orphaned nodes along with the edits that orphaned them or update references automatically along with entity renaming or moving edits, thereby resulting in a valid document at each point along the undo/redo history.

Luckily, all JavaScript runs on a single thred, with all events being processed in order. So as long as a plugin dispatches a series of edits without explicitly calling or yielding to other code in between, it can completely count on the document not being changed from under it in between edits. JavaScript functions thereby behave as transactions when it comes to our `Edit` API.

Given these thoughts, do you agree that the current solution of "squash with previous" is good enough for the use cases that need it? If so, do you still see other use cases that require a more involved API to solve?

trusz commented:
I think a lot of the concept are thought through, but I can imagine if you are not familiar with the history some things could harder to understand. For example I don't know why there are no more validator plugins :) 

Do you think this is the document to add these reasoning or some other place?

ca-d commented on API.md:
```diff
@@ -0,0 +1,329 @@
+
+An **OpenSCD plugin** is a [JavaScript module](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Modules) which exports a particular [custom element class](https://developer.mozilla.org/en-US/docs/Web/API/Web_components/Using_custom_elements#implementing_a_custom_element) as its [default export](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Modules#default_exports_vers ...
+
+An **OpenSCD menu plugin** is an OpenSCD plugin with an additional `run()` instance method which is called when the plugin's entry is clicked in OpenSCD's menu. It is continuously rendered into the app and is expected to normally not display any content of its own. It is meant for one-shot editing tasks or tasks that always run in the background (like validation).
```
Yes, I suggest we implement validators in a menu plugin like the one in https://openenergytools.github.io/scl-editor/ .

Jakob Vogelsang just merged both current validators into one codebase for his schema+template validator plugin, but I think OMICRON may well implement a "validator orchestration" plugin that loads its own validating sub-plugins so we can still have extensible validation without having to implement the API in core.

Do you see an additional need for a separate plugin type implemented in core just for validators?

ca-d commented:
> Do you think this is the document to add these reasoning or some other place?

I think an extra "History" section in the `README.md` might be a great idea, too. Do you think it would be better to have this section in the `API.md`?

trusz commented:
You could also document the individual decisions as ADR and link them here ;)

jarradraumati commented on API.md:
````diff
@@ -0,0 +1,335 @@
+```typescript
+/* eslint-disable no-undef */
+interface WizardRequestBase {
+  subWizard?: boolean;
````
I would be interested to hear other's thoughts on how `subWizard` was intended to be used.
I tried an approach for an `lnode` wizard for mapping `LN`s onto a `LNode` within the substation section of the scd file.

https://github.com/OpenEnergyTools/scl-wizarding/pull/10/files#diff-7b8cd3a9c9d707212e2cb17c3aabdabf7e0e0de291f3b510a01c42b519fac90e

This PR switches to a slightly different version of the `lNode` wizard.

Was this the intended use of this feature?

ca-d replied:
Thanks for your input into this, Jarrad! I'm glad to see you interested in the future of this API.

The intended use is when you want to jump to a different view to let the user edit some detail or dependent piece of data but want to jump back to the currently opened wizard as soon as they have saved or cancelled their changes. This allows us, for example, to jump to editing `DO`s from `LNodeTypes`, and then go back to continuing to edit the same `LNodeType`.

JakobVogelsang replied:
Sorry for having you waited so long on the linked review. I will get at it today and will leave comments in the review how `subWizard` works.

jarradraumati commented on API.md:
```diff
@@ -0,0 +1,335 @@
+
+The combination of `name` and `kind` uniquely identifies the plugin to be configured. If `config` is `null`, the plugin is removed. Otherwise, the plugin is added or reconfigured.
+
+## Theming
```
I have done some experimenting with Material 3 (in particular the latest version of [material-web](https://github.com/material-components/material-web) which takes a different approach to theming than what was done in Material 2.

One of the nice features was the way it dealt with light and dark modes (me being a fan of dark mode).
I am curious to know if there was any consideration for migrating to Material 3 (where possible using offical components, noting the team have not managed to complete all of them before going into maintenance mode)?
It was probably be a bit of refactoring amongst all of the plugins that are available.
Many of them at present are not really suited for dark mode.

ca-d replied:
Sounds like a great idea, but I wouldn't include a technical detail about the implementation in these API docs themselves. Let's open an issue for migrating to Material 3!

JakobVogelsang reviewed (approved):
LGTM

danyill reviewed (approved):
LGTM, thanks for all the discussion, 🎯 
Onward to implementation 🚀

JakobVogelsang reviewed (approved):
LGTM

jarradraumati reviewed (approved):
LGTM
